doc/rationale.md

```diff
@@ -0,0 +1,84 @@
+# Configy rationale
+
+The following explains the rationale versus various decisions made in Configy.
+
+## Use cases, Internal vs External schema
+
+Configy was originally designed with an application-first mindset: a developer should
+design their config file based on a `struct` definition they chose, and would potentially
+have a slightly different configuration file depending on the problem domain (for example,
+the `@Key("name")` attribute).
+
+While this is still the primary use-case for Configy, it is also very convenient to
+write a validator for schema that are outside of the application's control,
+such as `dub.json`, `.gitlab-ci.yml` / GitHub actions file, or a replacement for
+a JSON schema.
+
+The first use case (when the schema is under the developer's control) is referred to
+as an "Internal schema", while the second use case (when the schema is outside of the
+developer's control) is referred to as an "External schema". While both are supported,
+and Configy aims to have sane default that are likely to appear in external schemas,
+it might sometimes be required to "drop down" to using hooks (`fromString`, `fromYAML`)
+to accurately control deserialization for external schema.
+
+## Data vs Identity
+
+Configy being a DLang library adheres to one of the core tenet of the language,
+which is to have a clear separation between data (stored in `struct`) and objects
+with an identity (stored in `class`es). As configuration is data, Configy explicitly
+does not support classes.
+
+## The `struct` as the source of truth
+
+When designing with internal schema in mind, there should be a top-level `struct`
+that completely defines the program's configuration. Almost all of Configy's logic
+will be driven by this `struct`. It is not possible to have fields that are ignored
+by Configy - configuration shouldn't be mixed with non-configuration.
```

Comment:
I would also mention that we don't want to mix data with behaviour, so while struct methods are not technically forbidden, they are discouraged.